activemodel/lib/active_model/attribute_set.rb

```diff
@@ -104,7 +104,7 @@ def reverse_merge!(target_attributes)
     end
 
     def ==(other)
-      attributes == other.attributes
+      other.is_a?(AttributeSet.class) && attributes == other.send(:attributes)
```

Comment:
I'm sorry, I'm a bit confused. Should this really be `AttributeSet.class`?

Shouldn't it be
```suggestion
      other.is_a?(AttributeSet) && attributes == other.send(:attributes)
```
Or

```suggestion
      other.is_a?(self.class) && attributes == other.send(:attributes)
```

Not 100% sure this will work:

```suggestion
      other.is_a?(class) && attributes == other.send(:attributes)
```

Reply:
Looking at the discussion I think we want
```suggestion
      (other.is_a?(AttributeSet) || other.is_a?(LazyAttributeSet)) && attributes == other.send(:attributes)
```

Reply:
How about the following code? Actually {Lazy,}AttributeSet behaves a hash of attribute objects, like ActiveRecord::Relation behaves Array (Enumerable) like.

```suggestion
      if other.is_a?(Hash)
        attributes == other
      else
        other == attributes
      end
```

https://github.com/rails/rails/blob/00dfa109d6a5f4fb13d745f83d5cc779eba3b352/activerecord/lib/active_record/relation.rb#L797-L807

Reply:
My goal is to make `==(other)` safe without any change in behaviour.

So `other.is_a?(self.class) && attributes == other.send(:attributes)` won't work, because in this case `my_lazy_attribute_set == my_attribute_set` returns false. But in the original it returned true.

A similar way to ActiveRecord::Relation also introduces behavioural changes. So I'm not sure that such an implementation is the best fit here.

So I think `other.is_a?(AttributeSet) && attributes == other.send(:attributes)` would be the best option here. @MatheusRich thank you for the point about `.class`! I missed it indeed.